### ISSUE TITLE
 invaders + galaxian high scores get saved but not read back in (v0.158) . Other roms do work

### ISSUE THREAD

**Issue by beemandev** [libretro/mame#452]
UPDATE: 
Having done further testing with other roms (apart from invaders) I discovered that puckman, 1941, ckong DO work with high scores in Retroarch mame (latest) core whereas both invaders and galaxian high scores get saved but are not read back in when games are re-run (bad luck I started testing solely with invaders!)

Note, all roms, high scores work correctly using standalone Mame 0.158 + 0.158 merged rom set.

I deduced you also need as a minimum in plugins folder the json folder( along with  hiscore folder + boot.lua)

--original  post
I've spent about 12+ hours trying to track down why high scores don't get reloaded via mame core, when they do work fine in standalone mame (tested using game=invaders.zip)

I've devoured endless guides, posts, links and experimented with so many settings to no avail. So can only conclude that it's broken. I know I could workaround using save states via autosave option but for now I'll revert to using fbneo, but would much prefer to use Mame.

**Comment by sonninnos** [comment]
I've had zero issues using the hiscore plugin..

**Comment by beemandev** [comment]
Here's my notes on setting it all up (anything obvious missing?) :-

.Run Retroarch
.Main Menu/Load Core/Arcade (MAME)
.Start Core (without rom)
.Bring up Retroach Quick Menu/Core Options/System/Read Configuration=Enabled + Write Configuration=Disabled + Mame Ini Paths=Disabled 
.Resume content
.Mame Ui/(Tab)General Settings/Save Settings
.Mame Ui/(Tab)System Settings/Save System Settings + Advanced Options\Automatic save/restore=off
.Resume content
.Exit Mame OSD
.From original MAME folder copy plugins\hiscore folder + boot.lua to system\mame\plugins\
	copy plugin.ini to system\mame\ini\ ensuring 'hiscore	  1' is set
.Check system\mame\ini\mame.ini 'plugins	1' + 'autosave    0'
.Delete C:\Emu\RetroArch\saves\Arcade\mame\states\invaders\invaders\auto.sta
.RA/Settings/Saving/Load State Automatically=off

**Comment by beemandev** [comment]
And here is C:\Emu\RetroArch\config\MAME\MAME.opt
```
mame_alternate_renderer = "disabled"
mame_altres = "1280x960"
mame_auto_save = "disabled"
mame_autoloadfastforward = "disabled"
mame_boot_from_cli = "enabled"
mame_boot_to_bios = "disabled"
mame_boot_to_osd = "disabled"
mame_buttons_profiles = "disabled"
mame_cheats_enable = "disabled"
mame_cpu_overclock = "default"
mame_joystick_deadzone = "0.15"
mame_joystick_saturation = "0.85"
mame_joystick_threshold = "0.30"
mame_lightgun_mode = "lightgun"
mame_lightgun_offscreen_mode = "free"
mame_mame_4way_enable = "disabled"
mame_mame_paths_enable = "disabled"
mame_media_type = "rom"
mame_mouse_enable = "enabled"
mame_read_config = "enabled"
mame_rotation_mode = "libretro"
mame_saves = "game"
mame_softlists_auto_media = "enabled"
mame_softlists_enable = "enabled"
mame_thread_mode = "enabled"
mame_throttle = "disabled"
mame_write_config = "disabled"
```

**Comment by beemandev** [comment]
C:\Emu\RetroArch\system\mame\plugins\hiscore = 
```
hiscore.dat
init.lua
plugin.json
sort_hiscore.lua
```
C:\Emu\RetroArch\system\mame\plugins\boot.lua is present

C:\Emu\RetroArch\system\mame\ini\ui.ini  =
```
#
# UI SEARCH PATH OPTIONS
#
historypath               history;dats;.
categorypath              folders
cabinets_directory        cabinets;cabdevs
cpanels_directory         cpanel
pcbs_directory            pcb
flyers_directory          flyers
titles_directory          titles
ends_directory            ends
marquees_directory        marquees
artwork_preview_directory "artwork preview;artpreview"
bosses_directory          bosses
logos_directory           logo
scores_directory          scores
versus_directory          versus
gameover_directory        gameover
howto_directory           howto
select_directory          select
icons_directory           icons
covers_directory          covers
ui_path                   ui

#
# UI MISC OPTIONS
#
system_names              
skip_warnings             0
unthrottle_mute           0

#
# UI OPTIONS
#
infos_text_size           0.75
font_rows                 30
ui_border_color           ffffffff
ui_bg_color               ef101030
ui_clone_color            ff808080
ui_dipsw_color            ffffff00
ui_gfxviewer_color        ef101030
ui_mousedown_bg_color     b0606000
ui_mousedown_color        ffffff80
ui_mouseover_bg_color     70404000
ui_mouseover_color        ffffff80
ui_selected_bg_color      ef808000
ui_selected_color         ffffff00
ui_slider_color           ffffffff
ui_subitem_color          ffffffff
ui_text_bg_color          ef000000
ui_text_color             ffffffff
ui_unavail_color          ff404040

#
# SYSTEM/SOFTWARE SELECTION MENU OPTIONS
#
hide_main_panel           0
use_background            1
skip_biosmenu             0
skip_partsmenu            0
remember_last             1
last_used_machine         invaders
last_used_filter          Favorites
system_right_panel        image
software_right_panel      image
system_right_image        snap
software_right_image      snap
enlarge_snaps             1
forced4x3                 1
info_audit_enabled        0
hide_romless              1
```

C:\Emu\RetroArch\system\mame\ini\plugin.ini =
```
#
# PLUGINS OPTIONS
#
autofire                  0
cheat                     0
cheatfind                 0
console                   0
data                      1
discord                   0
dummy                     0
gdbstub                   0
hiscore                   1
inputmacro                0
layout                    0
portname                  0
timecode                  0
timer                     0
```

And AFAIK that's what's required. It works in the Mame standalone I copied files over from

**Comment by beemandev** [comment]
Can you confirm that both invaders.zip and galaxian.zip high scores work properly on your system?

I updated op with new info

**Comment by sonninnos** [comment]
Indeed, those games always show 0 even though something is saved. Pretty weird if they would work with the standalone, since this core does nothing different with plugins. The version is 0.258 though.

**Comment by beemandev** [comment]
Yes, it is weird, but definitely happens. I can rename/revert C:\Emu\Mame\hiscore\invaders.hi and see the high scores go and return with MAME standalone. 

re 0.258, yes I did all testing with standalone MAME v0.258 + RetroArch (both stable and nightly) + MAME (latest) core using MAME 0.258 merged roms from archive.org.

I tried other early games, asteroids high scores work fine, but headon doesn't save in RetroArch (Mame latest) or RetroArch (FBNeo) Or Standalone Mame. For that I'd have to turn autosave states on for just that rom I guess.

**Comment by grant2258** [comment]
Done a  little digging and found the frame notify is not getting triggered for lua. It could possibly be something missing in  the runloop. It doesnt run regardless of the high scores working or not. 

uncomment this https://github.com/libretro/mame/blob/3d612fb19eb95c0ae322c3cab343857b14a65a9c/plugins/hiscore/init.lua#L313C4-L313C4 (it shows up as not working better if you add a print comment as the top for the case of debugging) 

Games will fail to initialize at all if the memory test doesn't pass at the start as the recursive call isint working for lua.

https://github.com/libretro/mame/blob/3d612fb19eb95c0ae322c3cab343857b14a65a9c/plugins/hiscore/init.lua#L358C1-L361C6

@beemandev  some hi scores only save the tables so if you save a high score and check the tables, it might be the case that only the tables are saved for that game(headon).

edit confirmed its disabled
https://github.com/libretro/mame/blob/3d612fb19eb95c0ae322c3cab343857b14a65a9c/src/emu/video.cpp#L260-L263

**Comment by sonninnos** [comment]
That `MACHINE_NOTIFY_FRAME` is in `running_machine::retro_loop()` in `src/emu/machine.cpp` though. Found a few more where the scores aren't read back, or the save data is garbage since the file size is similarily small, such as "warpwarp" and "geebee".

The script indeed seems a bit weird, since a lot of stuff isn't even run (based on adding some verbose logging) when the hiscores are working properly..

So is the filesize the same with standalone, and does the same saved file work with standalone?

**Comment by grant2258** [comment]
The libretro core makes a pig ear of the galaxian hi sore that was saved on standalone, if thats what your asking. I haven't really looked into issue at all apart from the initial debugging the notify events arent happening. 

edit:
I would probably need to check more into how the output module connects up and if its related to that.  Im Pretty sure the standard one should work fine though. Ill try enable it and test when I have the time to compile mame again.

https://github.com/libretro/mame/blob/master/src/osd/modules/output/console.cpp

This is just pure guess work without tracing things properly.

**Comment by msheehan79** [comment]
This issue is still present on the 0.274 core, but I did find a potential workaround.  

While testing the lua script I found that adding a very small delay to the hiscore.dat entries for the affected games seems to solve the scores not getting read back in.  So my best guess is there is some sort of initial timing difference between standalone and the libretro version where its impacting the memory being ready to be changed by the plugin.

I have only tested invaders and galaxian so far, but in both cases adding a delay value of 0.1 seems to work.  I have not tested the other games in the same section to know if this is "safe" for every one of the games in the list, but at least its an option to get these games working with hiscores again.  I don't know how often the hiscore.dat changes these days, but the only potential downside I see is having to maintain a custom version of the dat with these delays added.

Here is what I put for invaders:

cosmicin:
galmonst:
invader4:
invaders:
invadersem:
invadrmr:
invasion:
invasiona:
invasiona2:
invasionb:
invasionrz:
invasionrza:
searthin:
searthina:
sicv:
sicv1:
sinvemag:
sinvzen:
sisv:
sisv1:
sisv2:
sisv3:
sitv:
sitv1:
spaceat2:
spaceatt:
spaceatt2k:
spaceattbp:
spacecom:
spacerng:
spceking:
superinv:
ultrainv:
@:maincpu,program,20f4,02,00,00
@delay=0.1

**Comment by sonninnos** [comment]
So is there harm just adding the delay always in the libretro version?

**Comment by grant2258** [comment]
This has to be an initialisation issue. Even the savestates are starting late to trigger them not working for rewind and fast forward dont work anymore. You would need to maintain a separate hi score for this core as a work around though. I havent looked at this core in a long time completely side tracked with other things.

**Comment by msheehan79** [comment]
I don't know if a blanket delay will impact some games negatively or not, unfortunately.  I kind of just stumbled across this by trial-and-error.  But, given this finding I also just tried adding the same 0.1 delay to strider2, and this seems to solve the black screen caused by the hiscore plugin on https://github.com/libretro/mame/issues/492.  So I agree seems like some timing/initialization issues going on that seem to manifest themselves differently depending on the game.

Possibly-related, I had also noticed on some games the hiscore plugin doesn't work only when runahead is enabled - an example of this is anteater.  Works fine without runahead but not when its enabled.  Tried adding a delay setting there again but no luck in that particular scenario.

**Comment by sonninnos** [comment]
Figured out how to remove the need for that delay, but doing so as is also adds one frame of input lag, so need to investigate more to see what is required to have no drawbacks.

All I did was test what happens when `MACHINE_NOTIFY_FRAME` is done in the usual place instead of the libretro hack loop.

Edit: Ok, next frame response and problem hiscores working with relatively harmless and tiny hackery, as in allow some first few frames to use the normal post-frame route, and then start using the pre-frame libretro route.
